### ISSUE TITLE
Feature: FastAP Loss Function

### ISSUE THREAD

**Issue by KarahanS** [qdrant/quaterion#200]
- [x] [FastAP Loss](https://openaccess.thecvf.com/content_CVPR_2019/papers/Cakir_Deep_Metric_Learning_to_Rank_CVPR_2019_paper.pdf) function will be implemented. 
- [x] Test for the new loss function will be implemented.

Reference implementations:
- https://github.com/kunhe/FastAP-metric-learning/blob/master/pytorch/FastAP_loss.py
- https://github.com/KevinMusgrave/pytorch-metric-learning/blob/691a6354be130547a8e26e170b86cf65c36cd791/src/pytorch_metric_learning/losses/fast_ap_loss.py

**Comment by KarahanS** [comment]
Please refer to the PR #199.

**Comment by KarahanS** [comment]
- FastAP loss function is implemented.
- Unittest is prepared - in which we compare the result of our loss function with that of the official implementation.
- Loss function is implemented to comply with `device` consistency and tested using gpu.
